app/src/main/java/com/quran/labs/androidquran/ui/TranslationManagerActivity.java

```diff
@@ -330,6 +356,86 @@ private void removeItem(final TranslationRowData translationRowData) {
     builder.show();
   }
 
+  private List<TranslationItem> sortedDownloadedItems() {
+    final ArrayList<TranslationItem> result = new ArrayList<>();
+    for (TranslationItem item : allItems) {
+      if (item.exists()) result.add(item);
+    }
+    Collections.sort(result, new TranslationItemDisplaySort ());
+    return result;
+  }
+
+  private void rankDownItem(TranslationRowData targetRow) {
+    TranslationItem targetItem = (TranslationItem) targetRow;
+    List<TranslationItem> sortedDownloads = sortedDownloadedItems();
+    if (sortedDownloads.indexOf(targetItem) + 1 < sortedDownloads.size()) { // ignore last item in list
+      TranslationItem updatedItem = targetItem.withDisplayOrder(targetItem.getDisplayOrder() + 1);
+      ArrayList<TranslationItem> toUpdate = new ArrayList<>();
+      toUpdate.add(updatedItem);
+      TranslationItem swapItem = null;
+      for(TranslationItem translationItem : sortedDownloads){
+        if (translationItem.getDisplayOrder() == updatedItem.getDisplayOrder()) {
+          swapItem = translationItem;
+          break;
+        }
+      }
+      if (swapItem != null) { // swap item order
+        if (swapItem.getDisplayOrder() > 0) {
+          toUpdate.add(swapItem.withDisplayOrder(swapItem.getDisplayOrder() - 1));
+        }
+      } else { // shift item order for higher items
+        for (TranslationItem translationItem : sortedDownloads) {
+          if (translationItem.getDisplayOrder() > -1 && translationItem.getDisplayOrder() < updatedItem.getDisplayOrder()) {
+            toUpdate.add(translationItem.withDisplayOrder(translationItem.getDisplayOrder() + 1));
+          }
+        }
+      }
+      if (!toUpdate.isEmpty()) {
+        if (selectionListener != null) selectionListener.handleSelection(updatedItem);
+        for(TranslationItem toUpdateItem : toUpdate){
+          updateTranslationItem(toUpdateItem);
+        }
+        generateListItems();
+      }
```

Comment:
this is expensive because every move is a database query when it doesn't need to be. we should just update when action mode is closed.

Reply:
Hm, `updateTranslationItem` updates both our activity data and database in one go.  But, yea, it's probably not necessary... then we'll have to break up `updateTranslationItem` for use in moving translations.